lib/howl/ui/command_line.moon

```diff
@@ -0,0 +1,503 @@
+    if @spillover
+      @spillover = @spillover .. text
+    else
+      @spillover = text
+
+  pop_spillover: =>
+    spillover = @spillover
+    @spillover = nil
+    return spillover or ''
+
+  post_keypress: =>
+    @enforce_left_pos!
+
+  on_update: =>
+    -- only call on_update() after run() ends and before finish() begins
+    return if not @current or @current.state != 'running'
+    if @_activity and @_activity.on_update and not @_updating
+      -- avoid recursive calls
+      @_updating = true
+      ok, err = pcall ->
+        @_activity\on_update @text
+      @_updating = false
+      if not ok
+        error err
+
+    @_adjust_height_rows!
+
+  enforce_left_pos: =>
+    -- don't allow cursor to go left into prompt
+    return if not @current
+    left_pos = @_prompt_end or 1
+    if @command_widget.cursor.pos < left_pos
+      @command_widget.cursor.pos = left_pos
+
+  handle_keypress: (event) =>
+    -- keymaps checked in order:
+    --   @preemptive_keymap - keys that cannot be remapped by activitys
+    --   @_activity.keymap
+    --   widget.keymap for widget in @_widgets
+    --   @keymap
+    @clear_notification!
+    @window.status\clear!
+
+    return true if bindings.dispatch event, 'commandline', { @preemptive_keymap}, self
+
+    activity = @_activity
+    if activity and activity.keymap
+      return true if bindings.dispatch event, 'commandline', { activity.keymap }, activity
+
+    if @_widgets
+      for _, widget in pairs @_widgets
+        if widget.keymap
+          return true if bindings.dispatch event, 'commandline', { widget.keymap }, widget
+
+    return true if bindings.dispatch event, 'commandline', { @keymap }, self
+
+    return false
+
+  preemptive_keymap:
+    ctrl_shift_backspace: => @abort_all!
+
+  keymap:
+    binding_for:
+      ["cursor-home"]: => @command_widget.cursor.pos = @_prompt_end
+
+      ["cursor-line-end"]: => @command_widget.cursor.pos = @_prompt_end + @text.ulen
+
+      ["cursor-left"]: => @command_widget.cursor\left!
+
+      ["cursor-right"]: => @command_widget.cursor\right!
+
+      ["editor-delete-back"]: =>
+        -- don't backspace into prompt
+        return true if @command_widget.cursor.pos == @_prompt_end
+        @command_widget\delete_back!
+        @on_update!
+
+      ["editor-paste"]: =>
+        import clipboard from howl
+        if clipboard.current
+          @write clipboard.current.text
+
+  add_widget: (name, widget) =>
+    error('No widget provided', 2) if not widget
+
+    @remove_widget name
+
+    @box\pack_end widget\to_gobject!, false, 0, 0
+    @_widgets[name] = widget
+
+    widget\show!
+
+  remove_widget: (name) =>
+    widget = @_widgets[name]
+    @box\remove widget\to_gobject! if widget
+    @_widgets[name] = nil
+
+  get_widget: (name) => @_widgets[name]
+
+  show: =>
+    return if @showing
+    @last_focused = @window.focus if not @last_focused
+    @window\_remember_focus!
```

Comment:
Yes, I was going to add a comment to the effect that I realized this but it didn't happen :) Anyway, it would be great if we could change it now at any rate.